---
title: "BSA"
author: "Deng Zhaoguo"
format: html
self-contained: true
date: "`r Sys.Date()`"
toc: true       
toc-depth: 5
execute:
  eval: TRUE
  echo: TRUE
  warning: FALSE
  message: FALSE
---

Bulk segregatants analysis，测序两个群体，wt pool 和 mutant pool, 每个pool混池测序，可以不是单基因决定

# 读取数据，原始数据请使用 BSA.slurm生成，合并gVCF再genotype

```{r}
# Read allele depth file
var_file = "Combine.results_var_AD.table"

raw = read.delim(var_file, header = TRUE, colClasses = c("character", "numeric", "character","character"))

# strsplit A,B \t X,Y to A B X Y
mut_AD = matrix(as.numeric(unlist(strsplit(raw[, 3], ","))), ncol = 2, byrow = TRUE)
wt_AD = matrix(as.numeric(unlist(strsplit(raw[, 4], ","))), ncol = 2, byrow = TRUE)

colnames(mut_AD) = colnames(wt_AD) = c("ref_reads", "alt_reads")
TEMP = cbind(raw[, 1:2], mut_AD)
merged_data = cbind(TEMP, wt_AD)

colnames(merged_data) = c("CHROM", "POS", "ref_reads_mut", "alt_reads_mut","ref_reads_wt", "alt_reads_wt")
```

输入数据格式:

![](images/BSA example.jpg)

# 计算等位基因频率

```{r}
# Calculate allele frequency in mut and wt pools
mut_freq = merged_data$alt_reads_mut / (merged_data$ref_reads_mut + merged_data$alt_reads_mut)
wt_freq = merged_data$alt_reads_wt / (merged_data$ref_reads_wt + merged_data$alt_reads_wt)

# mutate 0
mut_freq[mut_freq == 0] <- 1e-6
wt_freq[wt_freq == 0] <- 1e-6

# Calculate LOD
LOD = log10(
    dbinom(merged_data$alt_reads_mut, size = merged_data$alt_reads_mut + merged_data$ref_reads_mut, prob = mut_freq) *
    dbinom(merged_data$alt_reads_wt, size = merged_data$alt_reads_wt + merged_data$ref_reads_wt, prob = wt_freq) /
    (dbinom(merged_data$alt_reads_mut, size = merged_data$alt_reads_mut + merged_data$ref_reads_mut, prob = (mut_freq + wt_freq) / 2) *
    dbinom(merged_data$alt_reads_wt, size = merged_data$alt_reads_wt + merged_data$ref_reads_wt, prob = (mut_freq + wt_freq) / 2))
)

# combine
result_table = cbind(merged_data[, c("CHROM", "POS")], LOD)
```

## smooth函数

```{r}
# smooth function
smooth_LOD <- function(position, LOD, window_size = 1e6, step_size = 5e5) {
  start_positions <- seq(min(position), max(position) - window_size, by = step_size)
  smoothed_pos <- numeric(length(start_positions))
  smoothed_LOD <- numeric(length(start_positions))
  
  for (i in seq_along(start_positions)) {
    window_start <- start_positions[i]
    window_end <- window_start + window_size
    in_window <- position >= window_start & position <= window_end
    smoothed_pos[i] <- (window_start + window_end) / 2
    smoothed_LOD[i] <- mean(LOD[in_window], na.rm = TRUE)
  }
  
  return(data.frame(POS = smoothed_pos, LOD = smoothed_LOD))
}
```

# 绘图，分染色体

```{r}
# plot
for (chr in unique(result_table$CHROM)) {
  chr_data = subset(result_table, CHROM == chr)
  smoothed_data = smooth_LOD(chr_data$POS, chr_data$LOD)
  
  #jpeg(paste("LOD_", chr, ".jpg", sep = ""))
  
  # set y-axis range 0-10  and x-axis Mb
  plot(chr_data$POS / 1e6, chr_data$LOD, type = "n", 
       xlab = "Position (Mb)", ylab = "LOD score", 
       main = paste("Chromosome", chr),
       ylim = c(-2, 15))  # 设置纵坐标范围
  
  # draw LOD
  lines(smoothed_data$POS / 1e6, smoothed_data$LOD, type = "l", col = "red", lwd = 2)
  
  # Threshold at LOD-3
  abline(h = 3, col = "red", lwd = 2, lty = 2)  # lty = 2 设置为虚线
  
  axis(2, at = seq(0, 10, by = 1), labels = seq(0, 10, by = 1))
  
  #dev.off()
}
```

## 合并在一张图上

```{r}
# combine for one image
jpeg("LOD_all_chromosomes.jpg", width = 2100*2, height = 2970*2, res = 300)  # A4, 300 dpi
  
  # set 6 × 5
  par(mfrow = c(6, 5), mar = c(4, 4, 2, 1))
  
  for (chr in unique(result_table$CHROM)) {
    chr_data = subset(result_table, CHROM == chr)
    smoothed_data = smooth_LOD(chr_data$POS, chr_data$LOD)
    
    plot(chr_data$POS / 1e6, chr_data$LOD, type = "n", 
         xlab = "Position (Mb)", ylab = "LOD score", 
         main = paste("Chromosome", chr), 
         ylim = c(-2, 15))  
    
    lines(smoothed_data$POS / 1e6, smoothed_data$LOD, type = "l", col = "red", lwd = 2)
    
    abline(h = 3, col = "red", lwd = 2, lty = 2)
    
    axis(2, at = seq(0, 10, by = 1), labels = seq(0, 10, by = 1))
  }
  
dev.off()

```
